File: src/henaxel/raytracer/materials/Perlin.java
package henaxel.raytracer.materials;

import henaxel.raytracer.utils.Utils;
import henaxel.raytracer.utils.Vec3;

public class Perlin {
    private Vec3[] ranVec = perlinGenerate();
    private int[] permX = perlinGeneratePerm();
    private int[] permY = perlinGeneratePerm();
    private int[] permZ = perlinGeneratePerm();
    
    
    public double noise(Vec3 pos) {
        double u = pos.x - Math.floor(pos.x);
        double v = pos.y - Math.floor(pos.y);
        double w = pos.z - Math.floor(pos.z);
        int i = (int)Math.floor(pos.x);
        int j = (int)Math.floor(pos.y);
        int k = (int)Math.floor(pos.z);
        Vec3[][][] c = new Vec3[2][2][2];
        for(int di = 0; di < 2; di++) {
            for (int dj = 0; dj < 2; dj++) {
                for (int dk = 0; dk < 2; dk++) {
                    c[di][dj][dk] = ranVec[permX[(i + di) & 255] ^ permY[(j + dj) & 255] ^ permZ[(k + dk) & 255]];
                }
            }
        }
        return perlinInterp(c, u, v, w);
    }
    public double noise(double u, double v) {
        int i = (int)Math.floor(u);
        int j = (int)Math.floor(v);
        int k = 0;
        Vec3[][][] c = new Vec3[2][2][2];
        for(int di = 0; di < 2; di++) {
            for (int dj = 0; dj < 2; dj++) {
                for (int dk = 0; dk < 2; dk++) {
                    c[di][dj][dk] = ranVec[permX[(i + di) & 255] ^ permY[(j + dj) & 255] ^ permZ[(k + dk) & 255]];
                }
            }
        }
        return perlinInterp(c, u, v, 0);
    }
    
    private static Vec3[] perlinGenerate() {
        Vec3[] p = new Vec3[256];
        for(int i = 0; i < 256; i++) {
            p[i] = new Vec3(-1 + 2*Math.random(), -1 + 2*Math.random(), -1 + 2*Math.random()).unitVector();
        }
        return p;
    }
    
    private static void permute(int[] p) {
        for(int i = p.length-1; i > 0; i--) {
            int target = (int)(Math.random()*(i+1));
            int tmp = p[i];
            p[i] = p[target];
            p[target] = tmp;
        }
    }
    
    private static int[] perlinGeneratePerm() {
        int[] p = new int[256];
        for(int i = 0; i < 256; i++) {
            p[i] = i;
        }
        permute(p);
        return p;
    }
    
    private static double perlinInterp(Vec3[][][] c, double u, double v, double w) {
        double uu = u*u*(3-2*u);
        double vv = v*v*(3-2*v);
        double ww = w*w*(3-2*w);
        double accum = 0;
        for(int i = 0; i < 2; i++)
            for(int j = 0; j < 2; j++)
                for(int k = 0; k < 2; k++) {
                    Vec3 weightV = new Vec3(u-i, v-j, w-k);
                    accum +=
                            (i*uu + (1-i)*(1-uu))*
                            (j*vv + (1-j)*(1-vv))*
                            (k*ww + (1-k)*(1-ww))*Vec3.dot(c[i][j][k], weightV);
                }
        return accum;
    }
    
    double turb(Vec3 p, int depth) {
        double accum = 0;
        Vec3 tmp = p;
        double weight = 1;
        for (int i = 0; i < depth; i++) {
            accum+= weight*noise(tmp);
            weight *= 0.5;
            tmp = Vec3.multiply(tmp, 2);
        }
        return Math.abs(accum);
    }
    double turb(double u, double v, int depth) {
        double accum = 0;
        double i = u;
        double j = v;
        double weight = 1;
        for (int k = 0; k < depth; k++) {
            accum+= weight*noise(i, j);
            weight *= 0.5;
            i *= 2;
            j *= 2;
        }
        return Math.abs(accum);
    }
}

File: src/henaxel/raytracer/materials/Perlin.kt
package henaxel.raytracer.materials

import henaxel.raytracer.utils.Vec3
import kotlin.math.abs
import kotlin.math.floor
import kotlin.random.Random


class Perlin {
    private val ranVec = perlinGenerate()
    private val permX = perlinGeneratePerm()
    private val permY = perlinGeneratePerm()
    private val permZ = perlinGeneratePerm()

    fun noise(pos: Vec3): Double {
        val u = pos.x - floor(pos.x)
        val v = pos.y - floor(pos.y)
        val w = pos.z - floor(pos.z)
        val i = floor(pos.x).toInt()
        val j = floor(pos.y).toInt()
        val k = floor(pos.z).toInt()
        val c: Array<Array<Array<Vec3>>> = Array<Array<Array<Vec3>>>(2) {
            Array<Array<Vec3>>(2) {
                arrayOf(
                    Vec3(0,0,0),
                    Vec3(0,0,0)
                )
            }
        }
        for (di in 0..1) {
            for (dj in 0..1) {
                for (dk in 0..1) {
                    c[di][dj][dk] =
                        ranVec.get(permX[i + di and 255] xor permY[j + dj and 255] xor permZ[k + dk and 255])
                }
            }
        }
        return Perlin.perlinInterp(c, u, v, w)
    }

    fun noise(u: Double, v: Double): Double {
        val i = floor(u).toInt()
        val j = floor(v).toInt()
        val k = 0
        val c = Array(2) {
            Array(2) {
                arrayOf(
                    Vec3(0,0,0),
                    Vec3(0,0,0)
                )
            }
        }
        for (di in 0..1) {
            for (dj in 0..1) {
                for (dk in 0..1) {
                    c[di][dj][dk] =
                        ranVec.get(permX[i + di and 255] xor permY[j + dj and 255] xor permZ[k + dk and 255])
                }
            }
        }
        return perlinInterp(c, u, v, 0.0)
    }

    fun turb(p: Vec3, depth: Int): Double {
        var accum = 0.0
        var tmp: Vec3 = p
        var weight = 1.0
        for (i in 0 until depth) {
            accum += weight * noise(tmp)
            weight *= 0.5
            tmp *= 2
        }
        return abs(accum)
    }

    fun turb(u: Double, v: Double, depth: Int): Double {
        var accum = 0.0
        var i = u
        var j = v
        var weight = 1.0
        for (k in 0 until depth) {
            accum += weight * noise(i, j)
            weight *= 0.5
            i *= 2.0
            j *= 2.0
        }
        return abs(accum)
    }

    companion object {

        fun permute(p: Array<Int>) {
            for (i in p.size - 1 downTo 1) {
                val target = (Random.nextDouble() * (i + 1)).toInt()
                val tmp = p[i]
                p[i] = p[target]
                p[target] = tmp
            }
        }

        fun perlinGenerate(): Array<Vec3> {
            val p = Array(256) { Vec3(0,0,0) }
            for (i in 0..255) {
                p[i] = Vec3(-1 + 2 * Random.nextDouble(), -1 + 2 * Random.nextDouble(), -1 + 2 * Random.nextDouble()).norm()
            }
            return p
        }

        fun perlinGeneratePerm(): Array<Int> {
            val p = Array(256) {0}
            for (i in 0..255) {
                p[i] = i
            }
            permute(p)
            return p
        }

        fun perlinInterp(c: Array<Array<Array<Vec3>>>, u: Double, v: Double, w: Double): Double {
            val uu = u * u * (3 - 2 * u)
            val vv = v * v * (3 - 2 * v)
            val ww = w * w * (3 - 2 * w)
            var accum = 0.0
            for (i in 0..1) for (j in 0..1) for (k in 0..1) {
                val weightV = Vec3(u - i, v - j, w - k)
                accum += (i * uu + (1 - i) * (1 - uu)) *
                        (j * vv + (1 - j) * (1 - vv)) *
                        (k * ww + (1 - k) * (1 - ww)) * (c[i][j][k] * weightV)
            }
            return accum
        }

    }
}